### PR TITLE
EnzymeLogic cleanup

### PR DESCRIPTION
**Pull request by tgymnich**
Split up CreatePrimalAndGradient and remove the fwd mode stuff

### CODE REVIEW

File: enzyme/Enzyme/CApi.cpp [hunk 1]
```diff
@@ -323,6 +323,31 @@ LLVMBasicBlockRef EnzymeGradientUtilsAllocationBlock(GradientUtils *gutils) {
   return wrap(gutils->inversionAllocs);
 }
 
+LLVMValueRef EnzymeCreateDual(EnzymeLogicRef Logic, LLVMValueRef todiff,
```
Comment (wsmoses) [inline enzyme/Enzyme/CApi.cpp:326]: Perhaps CreateForwardDiff ?  Also you probably don't need to pass a derivative mode here (unless we want to leave an option to expand a new vector forward mode).

File: enzyme/Enzyme/EnzymeLogic.cpp [hunk 2]
```diff
@@ -3390,6 +3389,349 @@ Function *EnzymeLogic::CreatePrimalAndGradient(
...
+    }
+  }
+
+  for (Argument &Arg : gutils->newFunc->args()) {
+    if (Arg.hasAttribute(Attribute::Returned))
+      Arg.removeAttr(Attribute::Returned);
+    if (Arg.hasAttribute(Attribute::StructRet))
+      Arg.removeAttr(Attribute::StructRet);
+  }
+  if (gutils->newFunc->hasFnAttribute(Attribute::OptimizeNone))
```
Comment (wsmoses) [inline enzyme/Enzyme/EnzymeLogic.cpp:3672]: While we're cleaning up, can we move the function attribute code into a common function used by all methods.

File: enzyme/Enzyme/CApi.cpp (hunk 1 above)
Comment (tgymnich) [reply enzyme/Enzyme/CApi.cpp]: done

File: enzyme/Enzyme/EnzymeLogic.cpp (hunk 2 above)
Comment (tgymnich) [reply enzyme/Enzyme/EnzymeLogic.cpp]: The function attribute code for CreateAugmentedPrimal is not quite the same, so ill leave it out.

Comment (wsmoses) [reply enzyme/Enzyme/EnzymeLogic.cpp]: How precisely do they differ?

**Comment by tgymnich** [comment]
Let's remove CacheAnalysis and others in a future PR. We should should probably also add mode assertions to `legalRecompute`

**Comment by wsmoses** [comment]
Should definitely still remove those mincut / etc parts, but can be subsequent PR.

**Review by wsmoses — APPROVED** [review]

**Pull request merged**